{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d0fd4564-f035-43a5-8c6b-17a5150fa61f",
   "metadata": {},
   "source": [
    "show attributes counts for both human and mouse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b83d02f4-db48-4e9b-9e7c-25e7b9ad8454",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "from bs4 import BeautifulSoup\n",
    "import pandas as pd\n",
    "import csv\n",
    "import re\n",
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1db0e7da-46e4-45ea-af86-3c285fb30201",
   "metadata": {},
   "outputs": [],
   "source": [
    "def import_GSE():\n",
    "    GREIN_data = pd.read_csv(\"data/GREIN_data.csv\")\n",
    "    GREIN_data = GREIN_data[GREIN_data.Species != 'Rattus norvegicus'] #drop brown rat\n",
    "    GREIN_human = GREIN_data[GREIN_data.Species == 'Homo sapiens']\n",
    "    GREIN_mouse = GREIN_data[GREIN_data.Species == 'Mus musculus']\n",
    "    \n",
    "    GSE_human = GREIN_human['GEO accession'].tolist()\n",
    "    GSE_mouse = GREIN_mouse['GEO accession'].tolist()\n",
    "    \n",
    "    return GSE_human, GSE_mouse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ee4e6ae6-a309-4dc3-9ac3-01bb81670a03",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_url(geo_id):\n",
    "    return f\"https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc={geo_id}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3ef994eb-d50e-4af8-b838-fd3191406dcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fetch_page(url):\n",
    "    try:\n",
    "        response = requests.get(url)\n",
    "        return response.text if response.status_code == 200 else f\"Failed to retrieve the page. Status code: {response.status_code}\"\n",
    "    except requests.exceptions.RequestException as e:\n",
    "        return f\"Error: {e}\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ee60ea3e-5aa8-49e1-a51f-bec077eb2cf8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def scrape_geo_data(geo_id):\n",
    "    url = get_url(geo_id)\n",
    "    page_content = fetch_page(url)\n",
    "    \n",
    "    if not isinstance(page_content, str):\n",
    "        return page_content\n",
    "\n",
    "    soup = BeautifulSoup(page_content, 'html.parser')\n",
    "    gsm_links = soup.find_all('a', href=lambda href: href and href.startswith('/geo/query/acc.cgi?acc=GSM'))\n",
    "    gsm_values = [link.text for link in gsm_links]\n",
    "\n",
    "    # Store GSM values with their corresponding GSE (geo_id) in a dictionary\n",
    "    gse_gsm_dict = {geo_id: gsm_values}\n",
    "\n",
    "    return gse_gsm_dict\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8e3ffa57-bc1c-4490-9c8c-ffcb72447da0",
   "metadata": {},
   "outputs": [],
   "source": [
    "GSEs_human = import_GSE()[0]\n",
    "#GSM_human = [scrape_geo_data(GSE) for GSE in GSEs_human[:10]]\n",
    "#GSM_human\n",
    "GSM_human = {}\n",
    "for GSE in GSEs_human[:10]:\n",
    "    data = scrape_geo_data(GSE)\n",
    "    if isinstance(data, dict):\n",
    "        GSM_human.update(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a29178ba-177f-4663-837a-071f85a1f849",
   "metadata": {},
   "outputs": [],
   "source": [
    "def scrape_characteristics(geo_id):\n",
    "    url = get_url(geo_id)\n",
    "    page_content = fetch_page(url)\n",
    "    \n",
    "    if not isinstance(page_content, str):\n",
    "        return page_content\n",
    "\n",
    "    soup = BeautifulSoup(page_content, 'html.parser')\n",
    "    characteristics_label = soup.find('td', text='Characteristics')\n",
    "\n",
    "    if characteristics_label:\n",
    "        characteristics_content = characteristics_label.find_next_sibling('td')\n",
    "        return str(characteristics_content)\n",
    "\n",
    "    return f\"Failed to find Characteristics for {geo_id}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2fc46524-d74a-41ed-94e5-7754af17d05f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_characteristics(input_str):\n",
    "    input_str = re.sub(r'<td[^>]*>', '', input_str)\n",
    "    pattern = r'(\\w+): ([^<]+)'\n",
    "    matches = re.findall(pattern, input_str)\n",
    "    \n",
    "    characteristics_dictionary = dict(matches)\n",
    "    \n",
    "    return characteristics_dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "1d7c16be-0b40-40f1-b755-2e1b807b4817",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['GSE100027', 'GSE100040', 'GSE100075', 'GSE100081', 'GSE100092', 'GSE100099', 'GSE100118', 'GSE100183', 'GSE100206', 'GSE100210'])"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GSM_human.keys()\n",
    "# GSM_human.values()\n",
    "# GSM_human.items()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "3e8cc6e6-784f-48fd-a999-5d9fede67944",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "characteristics_dictionary: {'strain': 'C57BL/6J', 'age': '3 months', 'genotype': 'wild type'}\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_341951/3920716366.py:9: DeprecationWarning: The 'text' argument to find()-type methods is deprecated. Use 'string' instead.\n",
      "  characteristics_label = soup.find('td', text='Characteristics')\n"
     ]
    }
   ],
   "source": [
    "geo_id = \"GSM2683998\"\n",
    "characteristics_string = scrape_characteristics(geo_id)\n",
    "characteristics_dictionary = extract_characteristics(characteristics_string)\n",
    "print(\"characteristics_dictionary:\", characteristics_dictionary)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "2196d0e4-a11a-423c-bdbf-41a244da60e9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_341951/3920716366.py:9: DeprecationWarning: The 'text' argument to find()-type methods is deprecated. Use 'string' instead.\n",
      "  characteristics_label = soup.find('td', text='Characteristics')\n"
     ]
    }
   ],
   "source": [
    "results = {}\n",
    "for key, value in GSM_human.items():\n",
    "    #print(f\"Key: {key}\")\n",
    "    # print(\"Values:\")\n",
    "    for GSM in value:\n",
    "        characteristics_string = scrape_characteristics(GSM)\n",
    "        characteristics_dictionary = extract_characteristics(characteristics_string)\n",
    "        results[GSM] = characteristics_dictionary\n",
    "        #print(f\"Characteristics for {GSM}: {characteristics_dictionary}\")\n",
    "    #print(\"--------------\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "6ac363dc-e979-49b7-b82e-5088c68aa6af",
   "metadata": {},
   "outputs": [],
   "source": [
    "characteristics_dict = {}\n",
    "\n",
    "for gsm_id, characteristics in results.items():\n",
    "    for gse_id, gsm_list in GSM_human.items():\n",
    "        if gsm_id in gsm_list:\n",
    "            if gse_id not in characteristics_dict:\n",
    "                characteristics_dict[gse_id] = {}\n",
    "            characteristics_dict[gse_id][gsm_id] = characteristics\n",
    "\n",
    "#print(characteristics_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "a70c8b29-41d4-4d7c-bdd8-aac2a9e59fa1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'tissue': 'Breast',\n",
       " 'type': 'Myoepithelial',\n",
       " 'transfection': 'dCas9 3a3l delta'}"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "characteristics_dict['GSE100027']['GSM2668081']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "b82c9251-0fa5-493a-91dc-9cf1269958e0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'tissue': 'Breast',\n",
       " 'type': 'Myoepithelial',\n",
       " 'transfection': 'dCas9 3a3l delta'}"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "characteristics_dict['GSE100027']['GSM2668082']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "df5c8632-d051-417f-aacc-4504794f1c02",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'GSM2668081': {'tissue': 'Breast',\n",
       "  'type': 'Myoepithelial',\n",
       "  'transfection': 'dCas9 3a3l delta'},\n",
       " 'GSM2668082': {'tissue': 'Breast',\n",
       "  'type': 'Myoepithelial',\n",
       "  'transfection': 'dCas9 3a3l delta'},\n",
       " 'GSM2668083': {'tissue': 'Breast',\n",
       "  'type': 'Myoepithelial',\n",
       "  'transfection': 'dCas9 3a3l delta'},\n",
       " 'GSM2668084': {'tissue': 'Breast',\n",
       "  'type': 'Myoepithelial',\n",
       "  'transfection': 'dCas9 3a3l'},\n",
       " 'GSM2668085': {'tissue': 'Breast',\n",
       "  'type': 'Myoepithelial',\n",
       "  'transfection': 'dCas9 3a3l'},\n",
       " 'GSM2668086': {'tissue': 'Breast',\n",
       "  'type': 'Myoepithelial',\n",
       "  'transfection': 'dCas9 3a3l'},\n",
       " 'GSM2668087': {'tissue': 'Breast',\n",
       "  'type': 'Myoepithelial',\n",
       "  'transfection': 'untransfected'},\n",
       " 'GSM2668088': {'tissue': 'Breast',\n",
       "  'type': 'Myoepithelial',\n",
       "  'transfection': 'untransfected'},\n",
       " 'GSM2668089': {'tissue': 'Breast',\n",
       "  'type': 'Myoepithelial',\n",
       "  'transfection': 'untransfected'}}"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "characteristics_dict['GSE100027']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "850b03cf-7957-42a3-b21a-20745b1516f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Store the dictionary in a JSON file\n",
    "with open('data/charateristics_human.json', 'w') as json_file:\n",
    "    json.dump(characteristics_dict, json_file)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "79fb5a3d-ab23-4019-9b5f-bac0eda511ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "def extract_label_content(soup, label_text):\n",
    "    label = soup.find('td', text=label_text)\n",
    "    if label:\n",
    "        content = label.find_next_sibling('td').get_text()\n",
    "    else:\n",
    "        content = f\"{label_text} not found on the page.\"\n",
    "    return content\n",
    "\n",
    "def scrape_metadata(geo_id):\n",
    "    url = f\"https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc={geo_id}\"\n",
    "\n",
    "    try:\n",
    "        response = requests.get(url)\n",
    "\n",
    "        if response.status_code == 200:\n",
    "            soup = BeautifulSoup(response.text, 'html.parser')\n",
    "            \n",
    "            metadata = {\n",
    "                'Title': extract_label_content(soup, 'Title'),\n",
    "                'Experiment type': extract_label_content(soup, 'Experiment type'),\n",
    "                'Organism': extract_label_content(soup, 'Organism'),\n",
    "                'Summary': extract_label_content(soup, 'Summary'),\n",
    "                'Overall design': extract_label_content(soup, 'Overall design'),\n",
    "                'SRA': extract_label_content(soup, 'SRA'),\n",
    "                'Samples': [] \n",
    "            }\n",
    "\n",
    "           # Find all <a> tags with href attributes containing \"GSM\"\n",
    "            gsm_links = soup.find_all('a', href=lambda href: href and href.startswith('/geo/query/acc.cgi?acc=GSM'))\n",
    "\n",
    "            gsm_values = [link.text for link in gsm_links]\n",
    "\n",
    "           \n",
    "            metadata['Samples'] = gsm_values\n",
    "            \n",
    "            platforms_label = soup.find('td', text=re.compile(r'Platforms \\(\\d+\\)'))\n",
    "            if platforms_label:\n",
    "                metadata['Platforms'] = platforms_label.find_next_sibling('td').get_text()\n",
    "            else:\n",
    "                metadata['Platforms'] = \"Platforms not found on the page.\"\n",
    "\n",
    "            return metadata\n",
    "\n",
    "        else:\n",
    "            return f\"Failed to retrieve the page. Status code: {response.status_code}\"\n",
    "\n",
    "    except requests.exceptions.RequestException as e:\n",
    "        return f\"Error: {e}\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "b033d47c-a509-4110-8671-46ba0e0059dc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "metadata: {'Title': 'Human TFIIH kinase CDK7 regulates transcription-associated epigenetic modification', 'Experiment type': 'Expression profiling by high throughput sequencingGenome binding/occupancy profiling by high throughput sequencing', 'Organism': 'Homo sapiens', 'Summary': 'CDK7 phosphorylates the RNA polymerase II (pol II) CTD and activates the P-TEFb- associated kinase, CDK9, but its regulatory roles remain obscure. Using human CDK7 analog-sensitive (CDK7as) cells, we observed reduced capping enzyme recruitment, increased pol II promoter-proximal pausing, and defective termination at gene 3\\'-ends upon CDK7 inhibition. We also found that CDK7 regulates chromatin modifications downstream of transcription start sites. H3K4me3 spreading was restricted at gene 5\\'-ends and H3K36me3 was displaced toward gene 3\\'-ends in CDK7as cells. Together, these results implicate a CDK7-dependent \"CTD code\" that regulates epigenetic marks in addition to RNA processing and pol II pausing.', 'Overall design': 'WT and analogue sensitive Cdk7as mutant cells were treated with the ATP analogue NM-PP1 that specifically inhibits the Cdk7as mutant kinase. Using ChIP-seq and RNA-seq we tested the effects of Cdk7 inactivation on pol II distribution along genes, CTD Ser2 and Ser5 phosphorylation, capping enzyme recruitment, histone H3K4 and H3K36 methylation and mRNA expression', 'SRA': 'SRP109292', 'Samples': ['GSM2670975', 'GSM2670976', 'GSM2670977', 'GSM2670978', 'GSM2670979', 'GSM2670980', 'GSM2670981', 'GSM2670982', 'GSM2670983', 'GSM2670984', 'GSM2670985', 'GSM2670986', 'GSM2670987', 'GSM2670988', 'GSM2670989', 'GSM2670990', 'GSM2670991', 'GSM2670992', 'GSM2670993', 'GSM2670994', 'GSM2670995', 'GSM2670996', 'GSM2670997', 'GSM2670998', 'GSM2670999', 'GSM2671000', 'GSM2671001', 'GSM2671002', 'GSM2671003', 'GSM2671004', 'GSM2671005', 'GSM2671006', 'GSM2722166', 'GSM2722167'], 'Platforms': 'GPL20301\\nIllumina HiSeq 4000 (Homo sapiens)\\n\\nGPL21290\\nIllumina HiSeq 3000 (Homo sapiens)\\n\\n'}\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_341951/2108478228.py:4: DeprecationWarning: The 'text' argument to find()-type methods is deprecated. Use 'string' instead.\n",
      "  label = soup.find('td', text=label_text)\n",
      "/tmp/ipykernel_341951/2108478228.py:38: DeprecationWarning: The 'text' argument to find()-type methods is deprecated. Use 'string' instead.\n",
      "  platforms_label = soup.find('td', text=re.compile(r'Platforms \\(\\d+\\)'))\n"
     ]
    }
   ],
   "source": [
    "geo_id = \"GSE100040\"\n",
    "metadata = scrape_metadata(geo_id)\n",
    "print(\"metadata:\", metadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "98d5cede-811f-42a2-8ac6-a9501e7cd51d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_341951/2108478228.py:4: DeprecationWarning: The 'text' argument to find()-type methods is deprecated. Use 'string' instead.\n",
      "  label = soup.find('td', text=label_text)\n",
      "/tmp/ipykernel_341951/2108478228.py:38: DeprecationWarning: The 'text' argument to find()-type methods is deprecated. Use 'string' instead.\n",
      "  platforms_label = soup.find('td', text=re.compile(r'Platforms \\(\\d+\\)'))\n"
     ]
    }
   ],
   "source": [
    "results = {}\n",
    "for GSE in GSEs_human:\n",
    "    metadata = scrape_metadata(GSE)\n",
    "\n",
    "    results[GSE] = metadata\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9d3b938-1835-4f6f-b969-f67e70fa05ba",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "env",
   "language": "python",
   "name": "env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
