{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d520e664-c70c-411e-bcd9-abd834babed2",
   "metadata": {},
   "source": [
    "# Question - answering task"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7a5b6092-09b6-4191-b043-3456ed57f39b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# from transformers import AutoModelForQuestionAnswering, AutoTokenizer, pipeline\n",
    "\n",
    "# model_name = \"deepset/roberta-base-squad2\"\n",
    "\n",
    "# # a) Get predictions\n",
    "# nlp = pipeline('question-answering', model=model_name, tokenizer=model_name)\n",
    "# QA_input = {\n",
    "#     'question': 'What was conducted?',\n",
    "#     'context': 'Cancer cells exploit the epithelial-to-mesenchymal transition (EMT) program toward metastasis. Cytoskeletal regulators are dually required in mesenchymal cells by promoting EMT-induced migration and sustaining the EMT program itself. In search for novel regulators of metastasis, we conducted an shRNA screen targeting a class of microtubule regulators, the plus-end tracking proteins (+TIPs). We show that the +TIP ACF7 is required for both the maintenance of EMT and to promote migration. We identified HectD1 as a potent E3 ubiquitin ligase mediating ACF7 degradation. Depletion of HectD1 robustly increases ACF7 protein levels and this is sufficient to enhance migration and EMT in cells, as well as facilitate metastasis in vivo. Ours results report the HectD1/ACF7 axis as a novel regulator of metastasis of breast cancer cells.'\n",
    "# }\n",
    "# res = nlp(QA_input)\n",
    "\n",
    "# # b) Load model & tokenizer\n",
    "# model = AutoModelForQuestionAnswering.from_pretrained(model_name)\n",
    "# tokenizer = AutoTokenizer.from_pretrained(model_name)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "15c8d1c8-a521-4d69-9d65-8a5773b04979",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'score': 0.19460827112197876,\n",
       " 'start': 293,\n",
       " 'end': 308,\n",
       " 'answer': 'an shRNA screen'}"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "63e2db08-06c2-46ce-8000-60c6b90a4d49",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/xbi/.conda/envs/env/lib/python3.11/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "from transformers import AutoModelForQuestionAnswering, AutoTokenizer, pipeline\n",
    "\n",
    "model_name = \"deepset/roberta-base-squad2\"\n",
    "\n",
    "nlp = pipeline('question-answering', model=model_name, tokenizer=model_name)\n",
    "QA_input = {\n",
    "    'question': 'Which sequencing data was used?',\n",
    "    \n",
    "    'context': 'BET proteins commonly activate cellular gene expression, yet inhibiting their recruitment paradoxically reactivates latent HIV-1 transcription. Here we identify the short isoform of BET family member BRD4 (BRD4S) as a novel corepressor of HIV-1 transcription. We found that BRD4S was enriched in chromatin fractions of latently infected T cells and was more rapidly displaced from chromatin upon BET inhibition than the long isoform. BET inhibition induced marked nucleosome remodeling at the latent HIV-1 promoter, which was dependent on the activity of BAF, a SWI/SNF chromatin-remodeling complex with known repressive functions in HIV-1 transcription. BRD4S directly bound BRG1, a catalytic subunit of BAF, via its bromo- and ET domains and was necessary for BRG1 recruitment to latent HIV-1 chromatin. Using ChIP-seq combined with ATAC-seq data, we found that the latent HIV-1 promoter phenotypically resembles endogenous LTR sequences, pointing to a select role of BRD4S:BRG1 complexes in genomic silencing of invasive retroelements.'\n",
    "}\n",
    "res = nlp(QA_input)\n",
    "\n",
    "# b) Load model & tokenizer\n",
    "model = AutoModelForQuestionAnswering.from_pretrained(model_name)\n",
    "tokenizer = AutoTokenizer.from_pretrained(model_name)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a2267fc2-aa64-485b-8d94-7bc1b15f8512",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'score': 0.22430936992168427,\n",
       " 'start': 812,\n",
       " 'end': 843,\n",
       " 'answer': 'ChIP-seq combined with ATAC-seq'}"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7c3d007-894b-4279-8e95-66304b5fb1dc",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "env",
   "language": "python",
   "name": "env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
